### PR TITLE
Add support for metadata webhooks to Asset Store

### PR DESCRIPTION
**Pull request by michal-hudy**
**Description**

Changes proposed in this pull request:

- Added support for extracting metadata
- Added filtering to webhooks
- Changed Asset status

**Related issue(s)**

#3327

### CODE REVIEW

File: components/asset-store-controller-manager/pkg/assethook/engine/metadata_engine.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	return current
+}
+
+func (*metadataEngine) toFiles(results map[string]*json.RawMessage) []File {
+	files := make([]File, 0, len(results))
+	for k, v := range results {
+		file := File{Name: k, Metadata: v}
+		files = append(files, file)
```
Comment (magicmatatjahu) [inline components/asset-store-controller-manager/pkg/assethook/engine/metadata_engine.go]: Maybe:
```go
files = append(files, File{Name: k, Metadata: v})
```

File: docs/asset-store/docs/06-03-clusterasset.md [hunk 2]
```diff
@@ -85,13 +92,20 @@ This table lists all possible parameters of a given resource together with their
...
 | **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the mutation webhook services. |
 | **spec.source.mutationwebhookservice.name** |    **NO**   | Provides the name of the mutation webhook service. |
-| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
 | **spec.source.mutationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
 | **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed metadata specific for a given mutation service and its functionality. |
+| **spec.source.mutationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
+| **spec.source.metadatawebhookservice** |    **NO**   | Provides specification of the metadata webhook services. |
+| **spec.source.metadatawebhookservice.name** |    **NO**   | Provides the name of the metadata webhook service. |
```
Comment (magicmatatjahu) [inline docs/asset-store/docs/06-03-clusterasset.md]: Required

File: docs/asset-store/docs/06-03-clusterasset.md [hunk 3]
```diff
@@ -85,13 +92,20 @@ This table lists all possible parameters of a given resource together with their
...
+| **spec.source.metadatawebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
```
Comment (magicmatatjahu) [inline docs/asset-store/docs/06-03-clusterasset.md]: Required

File: docs/asset-store/docs/06-03-clusterasset.md (hunk 2 above)
Comment (magicmatatjahu) [inline docs/asset-store/docs/06-03-clusterasset.md]: Required

File: docs/asset-store/docs/06-03-clusterasset.md (hunk 3 above)
Comment (magicmatatjahu) [inline docs/asset-store/docs/06-03-clusterasset.md]: Required

File: docs/asset-store/docs/06-01-asset.md [hunk 4]
```diff
@@ -88,13 +95,19 @@ This table lists all possible parameters of a given resource together with their
...
-| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
 | **spec.source.mutationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
-| **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed metadata specific for a given mutation service and its functionality. |
+| **spec.source.mutationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
+| **spec.source.metadatawebhookservice** |    **NO**   | Provides specification of the metadata webhook services. |
+| **spec.source.metadatawebhookservice.name** |    **NO**   | Provides the name of the metadata webhook service. |
+| **spec.source.metadatawebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
```
Comment (magicmatatjahu) [inline docs/asset-store/docs/06-01-asset.md]: Required

File: docs/asset-store/docs/06-01-asset.md [hunk 5]
```diff
@@ -88,13 +95,19 @@ This table lists all possible parameters of a given resource together with their
...
 | **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the mutation webhook services. |
 | **spec.source.mutationwebhookservice.name** |    **NO**   | Provides the name of the mutation webhook service. |
-| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
 | **spec.source.mutationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
-| **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed metadata specific for a given mutation service and its functionality. |
+| **spec.source.mutationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
+| **spec.source.metadatawebhookservice** |    **NO**   | Provides specification of the metadata webhook services. |
+| **spec.source.metadatawebhookservice.name** |    **NO**   | Provides the name of the metadata webhook service. |
```
Comment (magicmatatjahu) [inline docs/asset-store/docs/06-01-asset.md]: Required

File: docs/asset-store/docs/06-03-clusterasset.md [hunk 6]
```diff
@@ -85,13 +92,20 @@ This table lists all possible parameters of a given resource together with their
 | **spec.source.filter** |    **NO**   | Specifies the regex pattern used to select files to store from the package. |
 | **spec.source.validationwebhookservice** |    **NO**   | Provides specification of the validation webhook services. |
 | **spec.source.validationwebhookservice.name** |    **NO**   | Provides the name of the validation webhook service. |
```
Comment (magicmatatjahu) [inline docs/asset-store/docs/06-03-clusterasset.md]: Required

File: docs/asset-store/docs/06-03-clusterasset.md [hunk 7]
```diff
@@ -85,13 +92,20 @@ This table lists all possible parameters of a given resource together with their
...
-| **spec.source.validationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.validationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
```
Comment (magicmatatjahu) [inline docs/asset-store/docs/06-03-clusterasset.md]: Required

File: components/asset-store-controller-manager/pkg/assethook/engine/metadata_engine.go (hunk 1 above)
Comment (magicmatatjahu) [inline components/asset-store-controller-manager/pkg/assethook/engine/metadata_engine.go]:
```go
files = append(files, File{Name: k, Metadata: v})
```

File: components/asset-store-controller-manager/pkg/assethook/engine/metadata_engine.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+			return nil, errors.Wrap(err, "while sending request to metadata webhook")
+		}
+
+		results = e.mergeMetadata(results, response.Data)
+	}
+
+	return e.toFiles(results), nil
+}
+
+func (*metadataEngine) mergeMetadata(current map[string]*json.RawMessage, results []v1alpha1.MetadataResultSuccess) map[string]*json.RawMessage {
```
Comment (magicmatatjahu) [inline components/asset-store-controller-manager/pkg/assethook/engine/metadata_engine.go]: `mergeMetadata` is for my a bad name, better is `replaceMetadata`. By `merge` developer can thinking about concatenating old fields with new fields, not replacing.

File: docs/asset-store/docs/06-01-asset.md [hunk 9]
```diff
@@ -88,13 +95,19 @@ This table lists all possible parameters of a given resource together with their
...
+| **spec.source.metadatawebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
+| **spec.source.metadatawebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
+| **spec.source.metadatawebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
```
Comment (mmitoraj) [inline docs/asset-store/docs/06-01-asset.md]:
```suggestion
| **spec.source.metadatawebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files sent to the service. |
```

File: docs/asset-store/docs/06-03-clusterasset.md [hunk 10]
```diff
@@ -85,13 +92,20 @@ This table lists all possible parameters of a given resource together with their
...
 | **spec.source.validationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
+| **spec.source.validationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
```
Comment (mmitoraj) [inline docs/asset-store/docs/06-03-clusterasset.md]:
```suggestion
| **spec.source.validationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files sent to the service. |
```

File: docs/asset-store/docs/06-03-clusterasset.md [hunk 11]
```diff
@@ -85,13 +92,20 @@ This table lists all possible parameters of a given resource together with their
...
 | **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the mutation webhook services. |
 | **spec.source.mutationwebhookservice.name** |    **NO**   | Provides the name of the mutation webhook service. |
-| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
 | **spec.source.mutationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
 | **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed metadata specific for a given mutation service and its functionality. |
+| **spec.source.mutationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
```
Comment (mmitoraj) [inline docs/asset-store/docs/06-03-clusterasset.md]:
```suggestion
| **spec.source.mutationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files sent to the service. |
```

File: docs/asset-store/docs/06-03-clusterasset.md [hunk 12]
```diff
@@ -85,13 +92,20 @@ This table lists all possible parameters of a given resource together with their
...
+| **spec.source.metadatawebhookservice** |    **NO**   | Provides specification of the metadata webhook services. |
+| **spec.source.metadatawebhookservice.name** |    **NO**   | Provides the name of the metadata webhook service. |
+| **spec.source.metadatawebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
+| **spec.source.metadatawebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
+| **spec.source.metadatawebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
```
Comment (mmitoraj) [inline docs/asset-store/docs/06-03-clusterasset.md]:
```suggestion
| **spec.source.metadatawebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files sent to the service. |
```

File: docs/asset-store/docs/06-01-asset.md [hunk 13]
```diff
@@ -88,13 +95,19 @@ This table lists all possible parameters of a given resource together with their
...
 | **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the mutation webhook services. |
 | **spec.source.mutationwebhookservice.name** |    **NO**   | Provides the name of the mutation webhook service. |
-| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
 | **spec.source.mutationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
-| **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed metadata specific for a given mutation service and its functionality. |
+| **spec.source.mutationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
+| **spec.source.metadatawebhookservice** |    **NO**   | Provides specification of the metadata webhook services. |
```
Comment (mmitoraj) [inline docs/asset-store/docs/06-01-asset.md]:
```suggestion
| **spec.source.metadatawebhookservice** |    **NO**   | Provides specification of the metadata webhook service. |
```
Btw could you also correct for spec of the validation and mutation services? (lines 96 and 101)?

File: docs/asset-store/docs/06-03-clusterasset.md [hunk 14]
```diff
@@ -85,13 +92,20 @@ This table lists all possible parameters of a given resource together with their
...
 | **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the mutation webhook services. |
 | **spec.source.mutationwebhookservice.name** |    **NO**   | Provides the name of the mutation webhook service. |
-| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. |
 | **spec.source.mutationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
 | **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed metadata specific for a given mutation service and its functionality. |
+| **spec.source.mutationwebhookservice.filter** |    **NO**   | Specifies the regex pattern used to select files to send to the service. |
+| **spec.source.metadatawebhookservice** |    **NO**   | Provides specification of the metadata webhook services. |
```
Comment (mmitoraj) [inline docs/asset-store/docs/06-03-clusterasset.md]:
```suggestion
| **spec.source.metadatawebhookservice** |    **NO**   | Provides specification of the metadata webhook service. |
```
Btw could you also correct for spec of the validation and mutation services (lines 93 and 98)?

File: docs/asset-store/docs/06-01-asset.md (hunk 13 above)
Comment (michal-hudy) [reply docs/asset-store/docs/06-01-asset.md]: But this variable provided array od services

File: docs/asset-store/docs/06-03-clusterasset.md (hunk 14 above)
Comment (michal-hudy) [reply docs/asset-store/docs/06-03-clusterasset.md]: But this variable provided array od services

File: components/asset-store-controller-manager/pkg/handler/asset/asset.go [hunk 15]
```diff
@@ -254,6 +282,40 @@ func (h *assetHandler) onPending(ctx context.Context, object MetaAccessor, spec
 	return h.getReadyStatus(object, h.getBaseUrl(bucketStatus.URL, object.GetName()), files, pretty.Uploaded), nil
 }
 
+func (h *assetHandler) buildFiles(filenames []string) []v1alpha2.AssetFile {
```
Comment (pkosiec) [inline components/asset-store-controller-manager/pkg/handler/asset/asset.go]: populate Would be a better word

File: components/asset-store-controller-manager/pkg/handler/asset/asset.go [hunk 16]
```diff
@@ -254,6 +282,40 @@ func (h *assetHandler) onPending(ctx context.Context, object MetaAccessor, spec
...
+	result := make([]v1alpha2.AssetFile, 0, len(filenames))
+
+	for _, filename := range filenames {
+		result = append(result, v1alpha2.AssetFile{Name: filename})
+	}
+
+	return result
+}
+
+func (h *assetHandler) mergeMetadata(files []v1alpha2.AssetFile, metadatas []engine.File) []v1alpha2.AssetFile {
```
Comment (pkosiec) [inline components/asset-store-controller-manager/pkg/handler/asset/asset.go]: yup replace would be better

**Pull request merged**